### ISSUE TITLE
Allow `_` as variable name?

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#24]
It is idiomatic Python to use `_` as a dummy / throwaway variable. In Acton, it doesn't work:
```
kll@Boxy:~/terastream/acton/test$ ../actonc --root main pinger.act && ./pinger 

******************** Syntax error
pinger.act:25:11:
   |
25 |         _ = env.exit(1)
   |           ^^^
unexpected "= e"
expecting "def", ':', or comma
```

I tried to assign the return value from `env.exit()` as to get around #22 but trying to assign to `_` fails with the above error. I don't think it should!? I'll classify as bug for now. If this is per design we can change classification.

**Comment by plajjan** [comment]
FWIW, the idiomatic way to write the above thing, instead of assigning to a dummy variable, is this:
```
await async env.exit(1)
```

**Comment by plajjan** [comment]
Appears to be fixed, perhaps by #1020? Closing.

**Comment by nordlander** [comment]
I'm afraid #1020 didn't really fix this, as the original problem seems to have been with the parsing of _ (ambiguity between _ as an assignment pattern and _ as the unknown effect decorator on a def, to be precise). I've been able to fix the parser problem, but I'm still reopening the issue because...

**Comment by nordlander** [comment]
... the _ syntax is still treated as an ordinary identifier by the current compiler! This might of course work just fine in many situations, but it also forces all occurrences of _ within a binding scope to have the same type! For example, the sequence
```
_ = "x"
_ = True
```
is currently causing a type error because it's viewed by the type checker as just a variant of
```
_name = "x"
_name = True
```
As a remedy I'm adding special treatment of _ to the language, such that _ now stands for a wildcard pattern where every occurrence can have its own type. Moreover, _ is no longer valid as a function or parameter identifier, it is only accepted in places where the grammar expects a _pattern_ (as the left side of an assignment, the bindings of a for-loop, etc). A typical use-case is in nested patterns, like
```
(_, mid, _) = function_returning_a_triple(...)
```
With the upcoming fix, the first and third components of the above triple do not have to be of the same type anymore!

**Comment by nordlander** [comment]
Fixed by PR #1061 , closing.
